gunrock/app/SimpleTemplate/SimpleTemplate_enactor.cuh

```diff
@@ -0,0 +1,417 @@
+        // <todo> add the same options as in template parameters here, e.g.:
+        // | (((EnactorT::Problem::FLAG & Mark_Predecessors) != 0) ?
+        // Update_Predecessors : 0x0)
+        // </todo>
+        > BaseIterationLoop;
+
+    SimpleTemplateIterationLoop() : BaseIterationLoop() {}
+
+    /**
+     * @brief Core computation of SimpleTemplate, one iteration
+     * @param[in] peer_ Which GPU peers to work on, 0 means local
+     * \return cudaError_t error message(s), if any
+     */
+    cudaError_t Core(int peer_ = 0)
+    {
+        // --
+        // Alias variables
+        
+        auto &data_slice = this -> enactor ->
+            problem -> data_slices[this -> gpu_num][0];
+        
+        auto &enactor_slice = this -> enactor ->
+            enactor_slices[this -> gpu_num * this -> enactor -> num_gpus + peer_];
+        
+        auto &enactor_stats    = enactor_slice.enactor_stats;
+        auto &graph            = data_slice.sub_graph[0];
+        auto &frontier         = enactor_slice.frontier;
+        auto &oprtr_parameters = enactor_slice.oprtr_parameters;
+        auto &retval           = enactor_stats.retval;
+        auto &iteration        = enactor_stats.iteration;
+        
+        // <todo> add problem specific data alias here:
+        auto &degrees = data_slice.degrees;
+        auto &visited = data_slice.visited;
+        // </todo>
+        
+        // --
+        // Define operations
+
+        // advance operation
+        auto advance_op = [
+            // <todo> pass data to lambda
+            degrees,
+            visited
+            // </todo>
+        ] __host__ __device__ (
+            const VertexT &src, VertexT &dest, const SizeT &edge_id,
+            const VertexT &input_item, const SizeT &input_pos,
+            SizeT &output_pos) -> bool
+        {
+            // <todo> Implement advance operation
+                        
+            // Mark src and dest as visited
+            atomicMax(visited + src, 1);
+            auto dest_visited = atomicMax(visited + dest, 1);
+            
+            // Increment degree of src
+            atomicAdd(degrees + src, 1);
+            
+            // Add dest to queue if previously unsen
+            return dest_visited == 0;
+            
+            // </todo>
+        };
+
+        // filter operation
+        auto filter_op = [
+            // <todo> pass data to lambda
+            // </todo>
+        ] __host__ __device__ (
+            const VertexT &src, VertexT &dest, const SizeT &edge_id,
+            const VertexT &input_item, const SizeT &input_pos,
+            SizeT &output_pos) -> bool
+        {
+            // <todo> implement filter operation
+            return true;
+            // </todo>
+        };
+        
+        // --
+        // Run
+        
+        // <todo> some of this may need to be edited depending on algorithmic needs
+        // !! How much variation between apps is there in these calls?
```

Comment:
Not a lot of variance in the calls I think. Only thing that makes sense is optimizations like a fused kernel call instead of two separate ones.